Update BackupWalletAuto: rename existing target file.

marlengit opened the pull request:
If the target backup file already exists instead of skipping the backup the existing file is renamed with a suffix time.old and the backup is performed.

marlengit committed:
Update BackupWalletAuto: rename existing target file.

ftrader commented:
Code looks good. Here is a a patch for the automatic test which tests the creation of the .old file (on node2 only). Can you please run it and if it passes, integrate it into the PR? We don't want to be leaving behind code which isn't tested ;-)

P.S. on my box this test now takes about 250sec to run. I wonder what is so expensive in terms of what it does? Would be nice to reduce the runtime...

Plus, I noticed that we're concatenating strings with '/' in them instead of using os.path.join() which would be more portable. I'm not sure if that will be a problem for executing this test on other platforms, but it could be something we need to improve later.

```
diff --git a/qa/rpc-tests/walletbackupauto.py b/qa/rpc-tests/walletbackupauto.py
index 2fa74dc..ea77a86 100755
--- a/qa/rpc-tests/walletbackupauto.py
+++ b/qa/rpc-tests/walletbackupauto.py
@@ -36,6 +36,8 @@ and that the debug.log contains a conflict message
 """

 import os
+import fnmatch
+import hashlib
 from test_framework.test_framework import BitcoinTestFramework
 from test_framework.util import *
 from random import randint
@@ -221,7 +223,10 @@ class WalletBackupTest(BitcoinTestFramework):
         if os.path.isfile(node1backupfile): node1backupexists = 1
         else: logging.info("Error backup does not exist: %s"%(node1backupfile))

-        if os.path.isfile(node2backupfile): node2backupexists = 1
+        if os.path.isfile(node2backupfile):
+            node2backupexists = 1
+            # take MD5 for comparison to .old file in later test
+            node2backupfile_orig_md5 = hashlib.md5(open(node2backupfile, 'rb').read()).hexdigest()
         else: logging.info("Error backup does not exist: %s"%(node2backupfile))

         if os.path.isfile(node3backupfile): node3backupexists = 1
@@ -297,5 +302,36 @@ class WalletBackupTest(BitcoinTestFramework):
         assert_equal(0,node4backupexists)
         assert_greater_than(len(debugmsg_list),0)

+        # test that existing wallet backup is preserved
+        # rewind node 2's chain to before backupblock
+        logging.info("Rewinding node 2 to before auto backup test preservation of existing backup file")
+        self.nodes[2].invalidateblock(self.nodes[2].getblockhash(backupblock))
+        logging.info("Stopping all nodes")
+        self.stop_four()
+        logging.info("Restarting node 2")
+        self.nodes[2] = start_node(2, self.options.tmpdir,["-keypool=100", "-autobackupwalletpath=./newreldir/", "-autobackupblock=%s"%(backupblock) ])
+        # check that there is no .old yet (node 2 needs to generate a block to hit the height)
+        old_files_found=[]
+        for file in os.listdir(tmpdir + "/node2/regtest/newreldir"):
+            if fnmatch.fnmatch(file, "wallet.dat.auto.%s.bak.*.old" % (backupblock)):
+                logging.info("old file found: %s" % file)
+                old_files_found.append(file)
+        assert_equal(0, len(old_files_found))
+        # generate a block to hit the backup block height
+        # this should cause the existing backup to be saved to a timestamped .old copy
+        self.nodes[2].generate(1)
+        for file in os.listdir(tmpdir + "/node2/regtest/newreldir"):
+            if fnmatch.fnmatch(file, "*.old"):
+                old_files_found.append(file)
+        assert_equal(1, len(old_files_found))
+        # check that the contents of the .old match what we recorded earlier for node 2's backup
+        # (the file should just have been renamed)
+        logging.info("Checking .old file %s" % old_files_found[0])
+        assert_equal(node2backupfile_orig_md5,
+                     hashlib.md5(open(tmpdir + "/node2/regtest/newreldir/" + old_files_found[0], 'rb').read()).hexdigest())
+        logging.info("Checksum ok - shutting down")
+        stop_node(self.nodes[2], 2)
+        self.start_four()
+
 if __name__ == '__main__':
     WalletBackupTest().main()
```

marlengit committed:
Merge branch 'master' of ssh://github.com/BTCfork/hardfork_prototype_1_mvf-bu into fix/auto_wallet_backup

marlengit committed:
Update to test creation of the .old file.

marlengit commented:
I've added your patch and also replaced path definitions using os.path.join.

The function which seems to take the longest time is do_one_round() which has 6 calls to one_send() and calls the sync_mempools() function.

ftrader commented:
Thanks for updating the test.
